caikit/runtime/service_generation/protoable.py

```diff
@@ -0,0 +1,158 @@
+
+    Args:
+        signature: Dict[str, Type]
+            module signature of parameters and types
+    """
+    protoables = {}
+    log.debug("Building protoable signature for %s", signature)
+    for arg, arg_type in signature.items():
+        protoable_type = handle_protoables_in_union(arg_type)
+        if protoable_type:
+            protoables[arg] = protoable_type
+
+    return protoables
+
+
+def handle_protoables_in_union(arg_type: Type) -> Type:
+    """Handles various protoable arg types from a Union.
+
+    If arg_type is a union, then this will return the union back if all types in it are proto-able,
+    or the first proto-able arg type if the union has non-protoable arg types.
+
+    If arg_type is not a union nor protoable at all, this returns None.
+
+    Examples:
+    Union[protoable_type, non_protoable_type] -> protoable_type
+    Union[protoable_type_1, protoable_type_2] -> Union[protoable_type_1, protoable_type_2]
+    Union[protoable_type_1, protoable_type_2, non_protoable_type] -> protoable_type_1
+    """
+    if is_protoable_type(arg_type):
+        if typing.get_origin(arg_type) == Union:
+            union_protoables = [
+                union_val
+                for union_val in typing.get_args(arg_type)
+                if is_protoable_type(union_val)
+            ]
+            # if all are protoable, return the union (which will create a oneof)
+            if len(union_protoables) == len(typing.get_args(arg_type)):
+                return arg_type
+            # if there's only 1 protoable found, return that
+            if len(union_protoables) == 1:
+                return union_protoables[0]
+            # otherwise, try to get the data model objects in the Union
+            dm_types = [
+                arg
+                for arg in union_protoables
+                if inspect.isclass(arg) and issubclass(arg, DataBase)
+            ]
+            # if there are multiple, pick the first one
+            if len(dm_types) > 0:
+                log.debug2(
+                    "Picking first data model type %s in union protoables %s",
+                    dm_types,
+                    union_protoables,
+                )
+                return dm_types[0]
+            log.debug(
+                "Just picking first protoable type %s in union",
+                union_protoables[0],
+            )
+            return union_protoables[0]
+        return arg_type
+    log.debug("Skipping non-protoable argument type [%s]", arg_type)
+
+
+def extract_data_model_type_from_union(arg_type: Type) -> Type:
+    """Helper function that determines the right data model type to use from a Union"""
+
+    # Decompose this type using typing to determine if it's a useful typing hint
+    typing_origin = get_origin(arg_type)
+    typing_args = get_args(arg_type)
+
+    # If this is a data model type, no need to do anything
+    if isinstance(arg_type, type) and issubclass(arg_type, DataBase):
+        return arg_type
```

Comment:
NIT: above you used `reflection.is_class` for this. I kind of like that since it's a little more explicit, but it would probably be good to be consistent

Reply:
Also, it looks like this same check is repeated in a lot of places, so it might be worth consolidating it into a helper function